File: backend/src/middleware/rateLimit.ts
import rateLimit from "express-rate-limit";

/** Applied globally (see app.ts) — generous enough not to bother normal
 * usage, but caps abusive/scripted traffic hitting any endpoint. */
export const generalRateLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  limit: 300,
  standardHeaders: true,
  legacyHeaders: false,
});

/** Applied only to /auth/login and /auth/refresh (see auth.router.ts) — much
 * tighter than the general limiter, since these are the credential-stuffing /
 * brute-force / refresh-token-guessing surface. Keyed by IP (the package's
 * default). */
export const authRateLimiter = rateLimit({
  windowMs: 15 * 60 * 1000,
  limit: 10,
  standardHeaders: true,
  legacyHeaders: false,
  message: { error: "Too many attempts, please try again later." },
});

/** Applied only to POST /ai/chat (see ai.router.ts) — each call costs real
 * FabriXAI tokens/money, so this stays tight like authRateLimiter rather
 * than the generous general limiter, but keyed per-user rather than the
 * package's IP default isn't needed here (IP-based is fine — the general
 * limiter already caps abusive traffic at a higher volume; this just caps
 * spend). */
export const aiRateLimiter = rateLimit({
  windowMs: 15 * 60 * 1000,
  limit: 20,
  standardHeaders: true,
  legacyHeaders: false,
  message: { error: "Too many AI requests, please try again later." },
});

File: backend-py/app/middleware/rate_limit.py
"""Port of `backend/src/middleware/rateLimit.ts`'s three limiters, using
`slowapi` (a `flask-limiter`-style wrapper around `limits`, IP-keyed by
default — same default key as `express-rate-limit`).

- `general_limiter` — applied globally (see `app/main.py`), generous.
- `auth_limiter` — applied only to `/auth/login` and `/auth/refresh`.
- `ai_limiter` — applied only to `POST /ai/chat` (not wired to a router yet
  in this phase — TODO(phase N): AI assistant router).
"""

from __future__ import annotations

from slowapi import Limiter
from slowapi.util import get_remote_address

GENERAL_RATE_LIMIT = "300/15minutes"
AUTH_RATE_LIMIT = "10/15minutes"
AI_RATE_LIMIT = "20/15minutes"

AUTH_RATE_LIMIT_MESSAGE = {"error": "Too many attempts, please try again later."}
AI_RATE_LIMIT_MESSAGE = {"error": "Too many AI requests, please try again later."}

# One shared Limiter instance (slowapi's recommended pattern) — each route
# opts into a tighter limit via `@limiter.limit(...)`; the app-wide default
# below is the "general" limiter applied to everything else.
limiter = Limiter(key_func=get_remote_address, default_limits=[GENERAL_RATE_LIMIT])
